create endpoint to register new user

raisa16 opened the pull request:
create User, UserController, UserDto, UserRepository, UserResponse, UserService, UserServiceImpl, UserType, UserTypeService, UserTypeServiceImpl, UserTypeRepository and  UserTypeNotFoundException update RestExceptionHandler add script to insert usertypes.

raisa16 committed:
WIP:
 create User, UserController, UserDto, UserRepository, UserResponse, UserService, UserServiceImpl, UserType, UserTypeService, UserTypeServiceImpl, UserTypeRepository and  UserTypeNotFoundException update RestExceptionHandler.

raisa16 committed:
update UserController, UserType, add script to insert usertypes

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/UserController.java:
```diff
@@ -0,0 +1,30 @@
+@RestController
+public class UserController {
+
+    UserService userService;
```
No olvidar el modificador de acceso

raisa16 replied:
corregido

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/UserController.java:
```diff
@@ -0,0 +1,30 @@
+        this.userService = userServiceImpl;
+    }
+
+    @PostMapping(path = "/users")
```
El mapping del nombre del recurso tiene mejor lugar en la definición de la anotación a nivel de la clase con una anotación específica. De esta forma todos lo métodos dentro de la clase seguiran el mismo nombre de recurso `/users` y sólo es separar por el tipo de método HTTP y no repetirlo.

En un desarrollo futuro, se tendría de nuevo que poner el nombre del recurso, esto sería repetirlo. La practica del uso de estas anotaciones de Spring Web sugiere una forma diferente.

raisa16 replied:
corregido

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/responses/UserResponse.java:
```diff
@@ -0,0 +1,31 @@
+    private Long id;
+    private String email;
+    @JsonProperty("user_type")
+    private String userType;
```
Sería mejor hacer uso de una clase enum con los tipos

raisa16 replied:
entonces seria crear una clase enume como "UserTypes", ya que la clase que se mapea con la table se llama UserTypes?, este valor quedaria registrado en el campo "name" de la tabla user_types?

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/UserServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public UserResponse create(UserDto userDto) {
+        User user = new User();
```
Es una mejor practica crear la instancia una vez que se asegura que el user type existe. De esta forma creamos un objeto que será inicializado con todos sus valores

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/UserServiceImpl.java:
```diff
@@ -0,0 +1,58 @@
+        return toUserResponse(userCreated);
+    }
+
+    public UserResponse toUserResponse(User user) {
```
Este tipo de métodos que hacen un procesamiento interno puede ser privado para que sea nada más de uso exclusivo dentro de la clase

raisa16 replied:
corregido

elbarbaro commented on src/main/resources/migrations/V5.create_user.sql:
```diff
@@ -0,0 +1,25 @@
+created_at timestamp not null,
+updated_at timestamp not null,
+deleted_at timestamp
+)
```
Nos faltó un `;`

raisa16 replied:
si se agrega el ; aun  se ejecutaria el segundo script ?

raisa16 replied:
corregido

elbarbaro commented on src/main/resources/migrations/V6.insert_user_types.sql:
Las instrucciones de DML están bien. Estas pueden ser incluidas también en la misma migración donde se crean las tablas. Una vez que se define los esquemas luego el crear estos registros

elbarbaro commented on src/main/resources/migrations/V6.insert_user_types.sql:
```diff
@@ -0,0 +1,5 @@
+
+INSERT INTO user_types(name, description, status, created_at, updated_at, deleted_at) VALUES
+('admin','this user has been created with all permissions','ACTIVE','2023-11-22 13:17:38.338333','2023-11-22 13:17:38.338333','null'),
```
Las fechas se podrían poner de manera automática al momento de que se ejecuten las instrucciones. Es recomendable usar funciones incorporadas del SGBD para consultar la fecha.

raisa16 committed:
latest updates for user data

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/services/UserServiceImpl.java:
```diff
@@ -0,0 +1,61 @@
+    @Override
+    public UserResponse create(UserDto userDto) {
+
+        com.nurserygarden.ecommerceapp.repositories.entities.UserType userType = userTypeRepository.findById(userDto.getUserTypeId()).orElseThrow(UserTypeNotFoundException::new);
```
Estamos duplicando la lógica que está en el UserTypeServiceImpl que es la misma comprobación de la existencia del tipo de usuario. La recomendación es utilizar ese componente en UserServiceImpl

raisa16 replied:
Se remplazo la logica para hacer uso de la excistente en UserTypeServiceImpl

elbarbaro commented on src/main/resources/migrations/V5.create_user.sql:
```diff
@@ -0,0 +1,31 @@
+('ADMIN','this user has been created with all permissions','ACTIVE',CURRENT_TIMESTAMP,CURRENT_TIMESTAMP);
+
+INSERT INTO user_types(name, description, status, created_at, updated_at, deleted_at) VALUES
+('user','this user has basic permissions','ACTIVE',CURRENT_TIMESTAMP,CURRENT_TIMESTAMP);
```
Este tipo no existe. Para mas referencia está la enumeración que se creo de UserType

raisa16 replied:
El tipo fue remplazado por 'CLIENT' de acuerdo a los tipos de usuarios permitidos

elbarbaro commented on src/main/java/com/nurserygarden/ecommerceapp/controllers/responses/UserType.java:
```diff
@@ -0,0 +1,5 @@
+package com.nurserygarden.ecommerceapp.controllers.responses;
+
+public enum UserType {
```
UserTypeValue sería un nombre para esta clase enum, para que no tenga conflicto con la clase de entidad UserType

raisa16 replied:
El nombre se actualizo a UserTypeValue

raisa16 committed:
-Updated enum class name from UserType to UserTypeValue and refactor this in UserResponse and UserServiceImpl classes.
-Update constructor injection to  reuse Exception in UserTypeServiceImpl exception.
-Update V5.create_user.sql to use valid name according to existing enum types and complete sql statment with missing value for the last column.
